component/dialer/sockopt_windows.go

```diff
@@ -0,0 +1,64 @@
+package dialer
+
+import (
+	"encoding/binary"
+	"net"
+	"syscall"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+func setSocketOptions(network, address string, c syscall.RawConn, opts *Options) (err error) {
+	if opts == nil || !isTCPSocket(network) && !isUDPSocket(network) {
+		return
+	}
+
+	var innerErr error
+	err = c.Control(func(fd uintptr) {
+		host, _, _ := net.SplitHostPort(address)
+		ip := net.ParseIP(host)
+		if ip != nil && !ip.IsGlobalUnicast() {
+			return
+		}
+
+		if opts.InterfaceIndex == 0 && opts.InterfaceName != "" {
+			if iface, err := net.InterfaceByName(opts.InterfaceName); err == nil {
+				opts.InterfaceIndex = iface.Index
+			}
+		}
+
+		if opts.InterfaceIndex != 0 {
+			switch network {
+			case "tcp4", "udp4":
+				innerErr = bindSocketToInterface4(windows.Handle(fd), uint32(opts.InterfaceIndex))
+			case "tcp6", "udp6":
+				innerErr = bindSocketToInterface6(windows.Handle(fd), uint32(opts.InterfaceIndex))
+				if network == "udp6" && ip == nil {
+					// the underlying IP net maybe IPv4 even if the 'network' param is 'udp6',
+					// so we should bind socket to interface4 at the same time
+					innerErr = bindSocketToInterface4(windows.Handle(fd), uint32(opts.InterfaceIndex))
+				}
+			}
+		}
+	})
+
+	if innerErr != nil {
+		err = innerErr
+	}
+	return
+}
+
+func bindSocketToInterface4(handle windows.Handle, interfaceIndex uint32) error {
+	const IP_UNICAST_IF = 31
```

Comment:
Do you think we should put the const within the func or declare it globally?

Reply:
It seems more reasonable that move `IP_UNICAST_IF` to global.

Reply:
And how about `IP_UNICAST_IF6`? They're different param, right?

Reply:
I think we can use `IP_UNICAST_IF` for all of them, since the value is the same. I've done some research, I found the `IP_UNICAST_IF` param is unique. Here is the windows.rs example: https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/Networking/WinSock/constant.IP_UNICAST_IF.html  I didn't find `IP_UNICAST_IF6` or similar.

Reply:
There is a param for IPv6: `IPV6_UNICAST_IF`, see https://learn.microsoft.com/en-us/windows/win32/winsock/ipproto-ipv6-socket-options, nut I wonder if they have the same value.

Reply:
Oh sorry, It is true that `IPV6_UNICAST_IF` is for IPV6.  I also find it in windows.rs and the value is `31` as well.  I've updated the code.